File: src/lang/FastMath.java
package lang;

/**
 * High-performance, 32-bit polynomial math approximations and bitwise operations.
 * Replaces Lookup Tables (LUTs) to avoid L1 Cache Misses and JVM Math calls.
 * Uses the Bhaskara I approximation with an extra precision pass and Quake III invSqrt.
 */
public final class FastMath {

    public static final float PI = 3.1415927f;
    public static final float HALF_PI = 1.5707964f;
    public static final float TWO_PI = 6.2831855f;
    public static final float PI2 = 6.2831855f;
    private static final float INV_PI2 = 0.15915494f;
    public static final float EPSILON = 0.000002f;

    // Polynomial constants
    private static final float B = 1.27323954f;  // 4 / PI
    private static final float C = -0.40528473f; // -4 / (PI^2)
    private static final float P = 0.225f;       // Precision weight

    // Precalculated multipliers to convert without division
    public static final float DEG_TO_RAD = 0.0174532925f; // PI / 180.0f
    public static final float RAD_TO_DEG = 57.2957795f;   // 180.0f / PI

    private FastMath() {}

    /**
     * Branchless Absolute Value for floats using IEEE 754 bit masking.
     */
    public static float abs(float x) {
        return Float.intBitsToFloat(Float.floatToRawIntBits(x) & 0x7FFFFFFF);
    }

    /**
     * Branchless Absolute Value for integers.
     */
    public static int abs(int n) {
        return (n ^ (n >> 31)) - (n >> 31);
    }

    /**
     * Fast Rounding using the 16384 magic float trick.
     * Returns float to avoid cast-back latency in math expressions.
     */
    public static float round(float x) {
        return (float) ((int) (x + 16384.5f) - 16384);
    }

    /**
     * Fast Inverse Square Root (The Quake III 0x5f3759df magic constant).
     * Computes 1 / sqrt(x) with high precision in a single Newton-Raphson iteration.
     */
    public static float invSqrt(float x) {
        float halfx = 0.5f * x;
        int i = Float.floatToRawIntBits(x);
        i = 0x5f3759df - (i >> 1);
        x = Float.intBitsToFloat(i);
        x = x * (1.5f - (halfx * x * x));
        return x;
    }

    /**
     * Fast Sine approximation (Bhaskara I with precision pass).
     */
    public static float sin32(float x) {
        x = x - PI2 * round(x * INV_PI2);
        final float absX = abs(x);
        float y = B * x + C * x * absX;
        float absY = abs(y);
        return P * (y * absY - y) + y;
    }

    /**
     * Fast Cosine approximation.
     */
    public static float cos32(float x) {
        x = x + HALF_PI;
        x = x - PI2 * round(x * INV_PI2);
        float absX = abs(x);
        float y = B * x + C * x * absX;
        float absY = abs(y);
        return P * (y * absY - y) + y;
    }

    /**
     * Fast Tangent: Computes sine and cosine in a single pass.
     */
    public static float tan32(float x) {
        float xSin = x - PI2 * round(x * INV_PI2);
        float absXSin = abs(xSin);
        float ySin = B * xSin + C * xSin * absXSin;
        float sin = P * (ySin * abs(ySin) - ySin) + ySin;

        float xCos = x + HALF_PI;
        xCos = xCos - PI2 * round(xCos * INV_PI2);
        float absXCos = abs(xCos);
        float yCos = B * xCos + C * xCos * absXCos;
        float cos = P * (yCos * abs(yCos) - yCos) + yCos;

        return sin / cos;
    }

    public static float toRadians(float degrees) {
        return degrees * DEG_TO_RAD;
    }

    public static float toDegrees(float radians) {
        return radians * RAD_TO_DEG;
    }

    public static float pow(float base, float exponent) {
        return (float) Math.pow(base, exponent);
    }

    public static float clamp(float val, float min, float max) {
        return Math.clamp(val, min, max);
    }

    public static float cosFromSin(float sin, float angle) {
        int quadrant = sin >= 0 ? (angle >= 0 ? 1 : 4) : (angle >= 0 ? 2 : 3);
        float cosSquared = 1.0f - sin * sin;
        return (float) (Math.sqrt(cosSquared) * (quadrant == 1 || quadrant == 4 ? 1 : -1));
    }
}

File: src/lang/fastmath.c
#include "lang/fastmath.h"

#include <math.h>
#include <string.h>

// fastmath.c — FastMath port (Legacy: lang/FastMath.java). Pure functions.

static const float INV_PI2 = 0.15915494f;   // 1 / (2*PI)
static const float B = 1.27323954f;          // 4 / PI
static const float C = -0.40528473f;         // -4 / (PI^2)
static const float P = 0.225f;               // precision weight

static uint32_t raw_bits(float x) {
    uint32_t bits;
    memcpy(&bits, &x, sizeof(bits));
    return bits;
}

static float from_bits(uint32_t bits) {
    float x;
    memcpy(&x, &bits, sizeof(x));
    return x;
}

float FastMath_abs(float x) {
    return from_bits(raw_bits(x) & 0x7FFFFFFFu);
}

int32_t FastMath_absInt(int32_t n) {
    const uint32_t u = (uint32_t)n;
    const uint32_t mask = (uint32_t)(n >> 31);
    return (int32_t)((u ^ mask) - mask);
}

float FastMath_round(float x) {
    return (float)((int32_t)(x + 16384.5f) - 16384);
}

float FastMath_invSqrt(float x) {
    float half_x = 0.5f * x;
    uint32_t i = raw_bits(x);
    i = 0x5f3759dfu - (i >> 1);
    x = from_bits(i);
    x = x * (1.5f - (half_x * x * x));
    return x;
}

float FastMath_sin32(float x) {
    x = x - FastMath_TWO_PI * FastMath_round(x * INV_PI2);
    const float abs_x = FastMath_abs(x);
    float y = B * x + C * x * abs_x;
    const float abs_y = FastMath_abs(y);
    return P * (y * abs_y - y) + y;
}

float FastMath_cos32(float x) {
    x = x + FastMath_HALF_PI;
    x = x - FastMath_TWO_PI * FastMath_round(x * INV_PI2);
    float abs_x = FastMath_abs(x);
    float y = B * x + C * x * abs_x;
    float abs_y = FastMath_abs(y);
    return P * (y * abs_y - y) + y;
}

float FastMath_tan32(float x) {
    float x_sin = x - FastMath_TWO_PI * FastMath_round(x * INV_PI2);
    float abs_x_sin = FastMath_abs(x_sin);
    float y_sin = B * x_sin + C * x_sin * abs_x_sin;
    float sin = P * (y_sin * FastMath_abs(y_sin) - y_sin) + y_sin;

    float x_cos = x + FastMath_HALF_PI;
    x_cos = x_cos - FastMath_TWO_PI * FastMath_round(x_cos * INV_PI2);
    float abs_x_cos = FastMath_abs(x_cos);
    float y_cos = B * x_cos + C * x_cos * abs_x_cos;
    float cos = P * (y_cos * FastMath_abs(y_cos) - y_cos) + y_cos;

    return sin / cos;
}

float FastMath_toRadians(float degrees) {
    return degrees * FastMath_DEG_TO_RAD;
}

float FastMath_toDegrees(float radians) {
    return radians * FastMath_RAD_TO_DEG;
}

float FastMath_pow(float base, float exponent) {
    return powf(base, exponent);
}

float FastMath_clamp(float val, float min, float max) {
    if (val < min)
        return min;
    if (val > max)
        return max;
    return val;
}

float FastMath_cosFromSin(float sin, float angle) {
    int quadrant = sin >= 0 ? (angle >= 0 ? 1 : 4) : (angle >= 0 ? 2 : 3);
    float cos_sq = 1.0f - sin * sin;
    float root = sqrtf(cos_sq);
    return (quadrant == 1 || quadrant == 4) ? root : -root;
}